lib/k8s/conn/auth/azure.ex

```diff
@@ -0,0 +1,89 @@
+defmodule K8s.Conn.Auth.Azure do
+  @moduledoc """
+  `auth-provider` for azure
+  """
+  alias K8s.Conn.Error
+  alias K8s.Conn.RequestOptions
+
+  require Logger
+  @behaviour K8s.Conn.Auth
+
+  defstruct [:token]
+
+  @type t :: %__MODULE__{
+          token: String.t()
+        }
+
+  @impl true
+  @spec create(map, String.t()) :: {:ok, t} | :skip
+  def create(
+        %{
+          "auth-provider" => %{
+            "config" => %{
+              "access-token" => token,
+              "tenant-id" => tenant,
+              "expires-on" => expires_on,
+              "refresh-token" => refresh_token,
+              "client-id" => client_id,
+              "apiserver-id" => apiserver_id
+            },
+            "name" => "azure"
+          }
+        },
+        _
+      ) do
+    if DateTime.diff(DateTime.utc_now(), parse_expires(expires_on)) >= 0 do
+      Logger.info(
+        "Azure token expired, using refresh token get new access, this will stop working when refresh token expires"
+      )
+
+      {:ok, %__MODULE__{token: refresh_token(tenant, refresh_token, client_id, apiserver_id)}}
+    else
+      {:ok, %__MODULE__{token: token}}
+    end
+  end
+
+  def create(_, _), do: :skip
+
+  @spec parse_expires(String.t()) :: DateTime.t()
+  defp parse_expires(expires_on) do
+    case Integer.parse(expires_on) do
+      {expires_on, _} -> DateTime.from_unix!(expires_on)
+      :error -> DateTime.from_iso8601(expires_on)
+    end
+  end
+
+  @spec refresh_token(String.t(), String.t(), String.t(), String.t()) :: String.t()
+  defp refresh_token(tenant, refresh_token, client_id, _apiserver_id) do
+    payload =
+      URI.encode_query(%{
+        "client_id" => client_id,
+        "grant_type" => "refresh_token",
+        "refresh_token" => refresh_token
+      })
+
+    {:ok, res} =
+      K8s.Client.MintHTTPProvider.request(
+        :post,
+        URI.new!("https://login.microsoftonline.com/#{tenant}/oauth2/v2.0/token"),
+        payload,
+        %{
```

Comment:
I think the function expects:

* uppercase string as method ("POST")
* a list of tuples, e.g. [{"Content-Type", "application/x-www-form-urlencoded"}] as headers

Reply:
Looking at the spec 
```
  @callback request(
              method :: atom(),
              uri :: URI.t(),
              body :: binary,
              headers :: list(),
              http_opts :: keyword()
            ) :: response_t()
```

I think it is correct to use the post atom, but I will change the header to a list